pkg/controller/poolcoordinator/poolcoordinator_controller.go

```diff
@@ -0,0 +1,270 @@
+	on := o.(*corev1.Node)
+	nn := n.(*corev1.Node)
+	//klog.Infof("node update: %v\n", nn.Name)
+	opool, ook := utils.NodeNodepool(on)
+	npool, nok := utils.NodeNodepool(nn)
+	if !ook && !nok {
+		return
+	}
+	if !ook && nok {
+		GetController().nodepoolMap.Add(npool, nn.Name)
+	} else if ook && !nok {
+		GetController().nodepoolMap.Del(opool, on.Name)
+	} else {
+		if opool != npool {
+			GetController().nodepoolMap.Del(opool, on.Name)
+			GetController().nodepoolMap.Add(npool, nn.Name)
+		}
+	}
+}
+
+func NewController(kc client.Interface, informerFactory informers.SharedInformerFactory) *Controller {
+	ldc = NewLeaseDelegatedCounter()
+
+	ctl = &Controller{
+		client: kc,
+	}
+
+	ctl.nodeInformer = informerFactory.Core().V1().Nodes()
+	ctl.nodeSynced = ctl.nodeInformer.Informer().HasSynced
+	ctl.nodeLister = ctl.nodeInformer.Lister()
+	ctl.leaseInformer = informerFactory.Coordination().V1().Leases()
+	ctl.leaseSynced = ctl.leaseInformer.Informer().HasSynced
+	ctl.leaseLister = ctl.leaseInformer.Lister().Leases(corev1.NamespaceNodeLease)
+
+	return ctl
+}
+
+func GetController() *Controller {
+	klog.Infof("ctl: %v", ctl)
+	return ctl
+}
+
+func (nc *Controller) taintNodeNotSchedulable(name string) {
+	node, err := nc.nodeLister.Get(name)
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	taints := node.Spec.Taints
+	if utils.TaintKeyExists(taints, constant.NodeNotSchedulableTaint) {
+		return
+	}
+	nn := node.DeepCopy()
+	t := corev1.Taint{
+		Key:    constant.NodeNotSchedulableTaint,
+		Value:  "true",
+		Effect: corev1.TaintEffectNoSchedule,
+	}
+	nn.Spec.Taints = append(nn.Spec.Taints, t)
+	nn, err = nc.client.CoreV1().Nodes().Update(context.TODO(), nn, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Error(err)
+	}
+}
+
+func (nc *Controller) deTaintNodeNotSchedulable(name string) {
+	node, err := nc.nodeLister.Get(name)
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	taints := node.Spec.Taints
+	taints, deleted := utils.DeleteTaintsByKey(taints, constant.NodeNotSchedulableTaint)
+	if !deleted {
+		return
+	}
+	nn := node.DeepCopy()
+	nn.Spec.Taints = taints
+	nn, err = nc.client.CoreV1().Nodes().Update(context.TODO(), nn, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Error(err)
+	}
+}
+
+func (nc *Controller) Run(stopCH <-chan struct{}) {
+	if !cache.WaitForCacheSync(stopCH, nc.nodeSynced, nc.leaseSynced) {
+		klog.Error("sync poolcoordinator controller timeout")
+	}
+
+	klog.Info("create nodepool map")
+	nc.nodepoolMap = utils.NewNodepoolMap()
+	nl, err := nc.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Error(err)
+	}
+	nc.nodepoolMap.Sync(nl)
+
+	nc.nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    onNodeCreate,
+		UpdateFunc: onNodeUpdate,
+		DeleteFunc: onNodeDelete,
+	})
+	nc.leaseInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    onLeaseCreate,
+		UpdateFunc: onLeaseUpdate,
+		DeleteFunc: nil,
+	})
```

Comment:
How about moving these configurations to the `NewController` func

Reply:
updated